ascanrulesAlpha: FP fixes for time based NoSQL

itsdivyanshjain opened the pull request:
Hey zap team, divyansh from astra security here. This PR contains fix for false positive which raises based on time-based nosql injection. Implemented solution is based on this PR:https://github.com/zaproxy/zap-extensions/pull/4173/files

thc202 commented:
The commit needs to be signed-off. (Squash the commits.)
The changelog should be updated.

thc202 commented on addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/timing/OnlineSimpleLinearRegression.java:
```diff
@@ -0,0 +1,135 @@
+ * By convention, we fix correlation and slope at 1.0 and the intercept at 0.0 when insufficient
+ * data points (<2) have been added.
+ */
+public class OnlineSimpleLinearRegression {
```
It might be a good time to move these to `commonlib`.

thc202 commented:
Both `ascanrules` and `ascanrulesAlpha` need to depend on the newer version of `commonlib` (i.e. `>= 1.15.0`).
https://github.com/zaproxy/zap-extensions/blob/22cd4abfd257e1c7d13d5797b7a94d2fd6baf8c8/addOns/ascanrules/ascanrules.gradle.kts#L16-L17

thc202 commented on addOns/ascanrules/CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 ## [55] - 2023-06-06
 ### Changed
 - The Parameter Tamper Scan rule now includes example alert functionality for documentation generation purposes (Issue 6119)
+- Moved Timing utils to commonlib
```
- Note that the entries should be added under the `Unreleased` version. This one is old/released already.
- Use `Depend on newer version of Common Library add-on.` instead.

thc202 commented on addOns/ascanrulesAlpha/CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 ## Unreleased
 ### Changed
 - Maintenance changes.
+- Fix for time based nosql injection
```
```suggestion
- Depend on newer version of Common Library add-on.

### Fixed
- Fix time-based false positives in NoSQL Injection - MongoDB scan rule.
```

thc202 commented on addOns/commonlib/CHANGELOG.md:
```diff
@@ -18,6 +18,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 - Add info URL.
 - Constant for default number of threads.
+- Added timing utils for fixing FPs
```
This one needs to be moved to "Unreleased", can also be something like `Add utilities for time-based checks.`

thc202 commented on addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/OnlineSimpleLinearRegression.java:
```diff
@@ -17,7 +17,7 @@
-package org.zaproxy.zap.extension.ascanrules.timing;
+package org.zaproxy.addon.commonlib.timing;
 
 /**
```
Add the `@since` JavaDoc tag to the moved non-test classes, e.g. `@since 1.15.0`.

kingthorin commented:
It would be nice if we could get https://github.com/zaproxy/zap-extensions/pull/4316 over the finish line then this rebased on top of it.

thc202 commented:
Better the other way around, the changes in that PR are not correct and this PR addresses that (e.g. by not relying on previous requests sent).

kingthorin commented:
Oh okay! :grinning:

kingthorin commented on addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:
```diff
@@ -374,72 +355,77 @@ public void scan(HttpMessage msg, String param, String value) {
-                LOGGER.debug("Trying  with the value: {}", SLEEP_INJECTION[index][0]);
+                AtomicReference<HttpMessage> message = new AtomicReference<>();
+                paramValue = sleepPayload.replace("{0}", String.valueOf(timeSleepSeconds));
+                LOGGER.debug("Trying  with the value: {}", sleepPayload);
```
I know the problem was there before but the double space may as well be fixed :wink:

thc202 commented on addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:
```diff
@@ -374,72 +355,77 @@ public void scan(HttpMessage msg, String param, String value) {
             String phase = null;
```
This along with all other unused code should be removed.

thc202 replied:
This is no fully addressed (at least the following are not used: `SLEEP_SHORT_TIME`, `SLEEP_LONG_TIME`, `INSERT_SHORT_TIME`, `INSERT_LONG_TIME`, `aveRtt`, `getAveRtts(List)`, `rtt`, `hadTimeout`).

itsdivyanshjain commented:
@thc202 @kingthorin added suggested changes

thc202 commented:
The build is failing, needs a `spotlessApply`.

kingthorin commented:
The code currently tries to set variables that weren't declared.

itsdivyanshjain commented:
@thc202 @kingthorin any  update

thc202 commented on addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:
```diff
@@ -362,84 +331,82 @@ public void scan(HttpMessage msg, String param, String value) {
-            boolean hadTimeout = false;
-            String sleepValueToInj;
+            int index = 0;
+            String sleepPayload = SLEEP_INJECTION[index][0];
```
Shouldn't this be inside the loop? Otherwise it will always be the same.

thc202 replied:
Maybe worth adding a test that verifies the payloads are different.

kingthorin commented:
@itsdivyanshjain are you able to address the latest feedback from @thc202?

thc202 commented:
This is on their roadmap.

itsdivyanshjain commented:
@kingthorin @thc202 implemented required changes, with additional improvements

thc202 commented:
Build is failing (needs another `spotlessApply`).

kingthorin commented on addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:
```diff
@@ -362,92 +341,85 @@ public void scan(HttpMessage msg, String param, String value) {
+
+                            // send the request and retrieve the response
+                            sendAndReceive(timedMsg, false);
+                            return timedMsg.getTimeElapsedMillis() / 1000.0;
```
`TimeUnit.MILLISECONDS.toSeconds(timedMsg.getTimeElapsedMillis())`

kingthorin commented on addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:
```diff
@@ -362,92 +341,85 @@ public void scan(HttpMessage msg, String param, String value) {
+
+                    if (isInjectable) {
+                        // We Found IT!
+                        // First do logging
```
I'm fine with the found it comment (I guess) but the logging comment probably isn't needed, LOGGER.debug makes it pretty clear.

itsdivyanshjain commented:
any update?

kingthorin commented:
Yup, there's outstanding review issues to be addressed.

itsdivyanshjain commented:
> Yup, there's outstanding review issues to be addressed.

didn't get it. can you elaborate

kingthorin commented:
🤷‍♂️ it was two weeks ago.

If you’ve addressed all the review comments thus far then it’s just waiting on another review I guess.

itsdivyanshjain commented:
> 🤷‍♂️ it was two weeks ago.
> 
> If you’ve addressed all the review comments thus far then it’s just waiting on another review I guess.

yeah, i already addressed all the comments since I asked the update.

psiinon commented:
We rely on contributors telling us that as we dont know if a push addresses any or all of the issues raised 😁

kingthorin commented on addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:
```diff
@@ -70,7 +70,7 @@ public class MongoDbInjectionScanRule extends AbstractAppParamPlugin {
     // time-based attack detection will try to take less than the following number of seconds
     // note: detection of this length will generally only happen in the positive (detecting) case.
-    private static final double BLIND_SECONDS_LIMIT = 20.0;
+    private double BLIND_SECONDS_LIMIT;
```
If it's no longer a constant it shouldn't be full caps and it should be moved down to 114-124 so that it's grouped with similar vars.

kingthorin commented on addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CommandInjectionScanRule.java:
```diff
@@ -291,7 +291,7 @@ public class CommandInjectionScanRule extends AbstractAppParamPlugin {
 
     /** The number of seconds used in time-based attacks (i.e. sleep commands). */
-    private int timeSleepSeconds = DEFAULT_TIME_SLEEP_SEC;
+    private int MAX_DELAY = DEFAULT_TIME_SLEEP_SEC;
```
Only constants should use FULL CAPS.

kingthorin commented on addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:
```diff
@@ -140,6 +112,7 @@ public class MongoDbInjectionScanRule extends AbstractAppParamPlugin {
                     CommonAlertTag.OWASP_2021_A03_INJECTION,
                     CommonAlertTag.OWASP_2017_A01_INJECTION,
                     CommonAlertTag.WSTG_V42_INPV_05_SQLI);
+    private int MAX_DELAY = DEFAULT_TIME_SLEEP_SEC;
```
This isn't a constant, it shouldn't be FULL CAPS.

kingthorin commented on addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:
```diff
@@ -217,11 +190,27 @@ public Map<String, String> getAlertTags() {
+                BLIND_SECONDS_LIMIT = 20.0;
+            }
+        } catch (ConversionException e) {
+            LOGGER.debug(
+                    "Invalid value for '{}': {}",
+                    RULE_SLEEP_TIME,
+                    this.getConfig().getString(RULE_SLEEP_TIME));
```
Is this really meant to log the same value twice?

kingthorin replied:
Okay

kingthorin commented on addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:
```diff
@@ -217,11 +190,27 @@ public Map<String, String> getAlertTags() {
     @Override
     public void init() {
+        try {
+            MAX_DELAY = this.getConfig().getInt(RULE_SLEEP_TIME, DEFAULT_TIME_SLEEP_SEC);
+            // sum of first term & last term (using AP)
+            BLIND_SECONDS_LIMIT =
+                    (BLIND_REQUEST_LIMIT / 2.0)
+                                    * (2 * (MAX_DELAY - BLIND_REQUEST_LIMIT + 1)
+                                            + (BLIND_REQUEST_LIMIT - 1))
+                            + 1;
+            if (BLIND_SECONDS_LIMIT < 0) {
+                MAX_DELAY = DEFAULT_TIME_SLEEP_SEC;
+                BLIND_SECONDS_LIMIT = 20.0;
+            }
```
If they're changed after substantiation they're no longer CONSTANTS.

kingthorin commented on addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/TimingUtils.java:
```diff
@@ -54,6 +58,7 @@ public class TimingUtils {
      * @throws IOException if the RequestSender throws an IOException, it will bubble up here
      */
     public static boolean checkTimingDependence(
+            int MAX_DELAY,
```
This isn't a CONSTANT.

kingthorin commented on addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/TimingUtils.java:
```diff
@@ -78,7 +83,8 @@ public static boolean checkTimingDependence(
         int requestsMade = 0;
         double secondsLeft = secondsLimit;
-        int currentDelay = 1;
+        int currentDelay = MAX_DELAY - requestsLeft + 1;
```
I'm not saying this is wrong but maybe it needs a code comment or something? It caught my eye that delay is altered by number of requests 🤷‍♂️ I'm headed for a meeting and don't have time to dig 😜

kingthorin commented on addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CommandInjectionScanRule.java:
```diff
@@ -215,7 +215,6 @@ public class CommandInjectionScanRule extends AbstractAppParamPlugin {
     private static final int BLIND_REQUEST_LIMIT = 5;
     // time-based attack detection will try to take less than the following number of seconds
     // note: detection of this length will generally only happen in the positive (detecting) case.
-    private static final double BLIND_SECONDS_LIMIT = 20.0;
```
Remove the comments as well?

kingthorin replied:
Or relocate them.

kingthorin commented on addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:
```diff
@@ -70,7 +70,6 @@ public class MongoDbInjectionScanRule extends AbstractAppParamPlugin {
     private static final int BLIND_REQUEST_LIMIT = 5;
     // time-based attack detection will try to take less than the following number of seconds
     // note: detection of this length will generally only happen in the positive (detecting) case.
-    private double BLIND_SECONDS_LIMIT;
```
Remove or relocate the comments?

itsdivyanshjain replied:
removing it because timing utils contains doc string which is sufficient enough to get to know it's working

kingthorin commented on addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/TimingUtils.java:
```diff
@@ -83,9 +83,16 @@ public static boolean checkTimingDependence(
+        int currentDelay = maxDelay - requestsLeft + 1;
         double lastDelay = 0;
 
+        if (currentDelay < 0) {
```
`<=` ?

Would we ever want it to actually be zero?

itsdivyanshjain replied:
this would work fine. But, i agree in actual we don't need it to zero

itsdivyanshjain commented:
@kingthorin @thc202 looking forward to get this merged if all changes are fine.

kingthorin commented on addOns/ascanrules/CHANGELOG.md:
```diff
@@ -4,7 +4,8 @@ All notable changes to this add-on will be documented in this file.
 ## Unreleased
-
+### Changed
+- Depend on newer version of Common Library add-on.
```
```suggestion
- Depend on newer version of Common Library add-on.
    - Re-work Command Injection time based payloads/functionality to reduce false positives.
```

kingthorin commented on addOns/commonlib/CHANGELOG.md:
```diff
@@ -12,6 +12,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Add solution to 'Brute Forcing Log-in Credentials', 'Brute Forcing Session Identifiers' and 'Brute Forcing Directories and Files' vulnerabilities (Issue 8056).
 - Update vulnerabilities' CWE references to use HTTPS scheme.
+### Added
+- Add utilities for time-based checks
```
Add a period please.

kingthorin commented on addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/TimingUtils.java:
```diff
@@ -17,11 +17,15 @@
+/**
+ * Utility class to host time-based blind detection algorithms.
+ *
+ * @since 1.15.0
```
This PR should be rebased and adjusted accordingly. commonlib 1.18 was released last week.

The CHANGELOGs in the ascan rules add-ons likely need to be re-jigged due to releases as well.

kingthorin reviewed (commented):
I think this is the last of it for me. Thanks for staying on this and pushing us, it really is appreciated. (Both the code and the persistence).

kingthorin reviewed (commented):
Should be in the Unreleased section

kingthorin commented on addOns/commonlib/src/main/java/org/zaproxy/addon/commonlib/timing/OnlineSimpleLinearRegression.java:
```diff
@@ -32,6 +32,8 @@
  * By convention, we fix correlation and slope at 1.0 and the intercept at 0.0 when insufficient
  * data points (<2) have been added.
+ *
+ * @since 1.19.0
```
Sorry I wasn't clear. 1.18.0 is out so this will be upcoming 1.19.0.

itsdivyanshjain replied:
done

itsdivyanshjain commented:
already done :)

kingthorin reviewed (approved):
LGTM

itsdivyanshjain commented:
@thc202 any update

kingthorin commented:
@KNOXDEV I believe you added the Linear Regression component, do you have any feedback on this?

KNOXDEV commented:
> @KNOXDEV I believe you added the Linear Regression component, do you have any feedback on this?

This makes sense to me. Good sanity check. Also cool to see this is getting worked into other modules. Thanks for looping me in as always.

thc202 commented:
@itsdivyanshjain could you integrate the latest changes? (We can also take a look at that.)

itsdivyanshjain commented:
> @itsdivyanshjain could you integrate the latest changes? (We can also take a look at that.)

sure

itsdivyanshjain commented:
@thc202 @kingthorin , ready for review again

thc202 commented:
I didn't check the latest changes but the build is failing (needs `spottlesApply`).

itsdivyanshjain commented:
> I didn't check the latest changes but the build is failing (needs `spotlessApply`).

done

thc202 commented:
Could you extract a PR moving just the timing utils to commonlib? That would make things simpler and allow to unblock other PRs.

thc202 commented on addOns/ascanrules/CHANGELOG.md:
```diff
@@ -4,6 +4,9 @@ All notable changes to this add-on will be documented in this file.
 ## Unreleased
+### Changed
+- Depend on newer version of Common Library add-on.
+- Re-work Command Injection time based payloads/functionality to reduce false positives.
```
This is covered with the following fixed entry.

thc202 commented on addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/CommandInjectionScanRule.java:
```diff
@@ -288,7 +288,7 @@ public class CommandInjectionScanRule extends AbstractAppParamPlugin {
 
     /** The number of seconds used in time-based attacks (i.e. sleep commands). */
-    private int timeSleepSeconds = DEFAULT_TIME_SLEEP_SEC;
+    private int maxDelay = DEFAULT_TIME_SLEEP_SEC;
```
This rename is not really needed.

thc202 commented on addOns/commonlib/CHANGELOG.md:
```diff
@@ -5,6 +5,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 
 ## Unreleased
+### Added
+- Add utilities for time-based checks.
```
Add new line after.

thc202 commented on addOns/commonlib/src/test/timing/TimingUtilsUnitTest.java:
```diff
@@ -146,4 +146,4 @@ void shouldAlertForCorrelatedTimes() throws IOException {
         assertThat(result, is(true));
     }
-}
+}
```
Restore newline.

itsdivyanshjain commented:
@thc202 New changes are implemented & tested. Do review this again

thc202 commented on addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:
```diff
@@ -55,83 +59,46 @@ public class MongoDbInjectionScanRule extends AbstractAppParamPlugin {
-            INSERT_LONG_TIME = "INSERT_LONG_TIME";
-    private static final int SHORT_THRESHOLD = 1500;
-    private static final int LONG_THRESHOLD = 3000;
+    private static final int BLIND_REQUEST_LIMIT = 5;
+    private static final int DEFAULT_TIME_SLEEP_SEC = 10;
```
Do these need to be different from the other scan rule? (The name of the first constant could also exactly match the other.)

thc202 commented on addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:
```diff
@@ -140,6 +107,7 @@ public class MongoDbInjectionScanRule extends AbstractAppParamPlugin {
                     CommonAlertTag.OWASP_2021_A03_INJECTION,
                     CommonAlertTag.OWASP_2017_A01_INJECTION,
                     CommonAlertTag.WSTG_V42_INPV_05_SQLI);
+    private int maxDelay;
```
`timeSleepSeconds` to be consistent with the other rule.

thc202 commented on addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:
```diff
@@ -362,92 +331,83 @@ public void scan(HttpMessage msg, String param, String value) {
         // The $where clause executes associated JS function one time for each tuple --> sleep time
         // = interval * nTuples
         if (!isBingo && doTimedScan) {
+            String paramValue;
```
Delay declaration to line 348.

thc202 commented on addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRule.java:
```diff
@@ -362,92 +331,83 @@ public void scan(HttpMessage msg, String param, String value) {
-            String sleepValueToInj;
+            int index = 0;
             while (index < SLEEP_INJECTION.length) {
+                String sleepPayload = SLEEP_INJECTION[index];
```
Do after the stop check.

thc202 commented on addOns/ascanrulesAlpha/src/test/java/org/zaproxy/zap/extension/ascanrulesAlpha/MongoDbInjectionScanRuleUnitTest.java:
```diff
@@ -79,4 +92,76 @@ void shouldReturnExpectedMappings() {
+                            int sleepInput = Integer.parseInt(match.group(1));
+                            Thread.sleep(sleepInput * 1000L);
+                        } catch (InterruptedException ex) {
+                            fail("failed to sleep thread for time-based nosql injection");
```
`// Ignore` instead, fail here will not fail the test, just the request (in any case the thread will not be interrupted normally).

thc202 commented:
Thank you!
